### PR TITLE
[Lang] Customized struct support

### LINKED ISSUE #2614: [Lang] Support customized struct in a SNode container?
**Concisely describe the proposed feature**

Make Taichi support user-defined struct (similar to a `namedtuple`). E.g.

```py
st = ti.types.make_struct(a=ti.i32, b=ti.f32, ...)
f = ti.field(st, shape=(8, 8))

@ti.kernel
def run():
  for i in f:
    f[i].b = f[i].a
```

**Describe the solution you'd like (if any)**

Right now one can achieve the same thing by separately defining `a` and `b` in the above example. However, supporting named struct would make the code more intuitive.

### PR DESCRIPTION
**Pull request by victoriacity**
Related issue = close #2614

Adds customized struct support to Taichi fields, e.g.:
```python
f = ti.Struct.field({'a':ti.i32, 'b':ti.f32}, shape=(8,))
# or
# f = ti.Struct.field({'a':ti.i32, 'b':ti.f32})
# ti.root.dense(ti.i, 8).place(f)
@ti.kernel
def run():
  for i in f:
    f[i].a = i
    f[i].b = f[i].a
```

The implementation generally follows the Matrix and MatrixField class with similar API definitions, while members of a Struct can be only accessed using attribute names.

A CompoundType interface has been implemented to instantiate vector, matrices, and structs with pre-defined shapes or members. This allows for a custom struct to have vectors, matrices and nested structs as members. 

```python
vec3f = ti.types.vector(3, ti.f32)
ray3f = ti.types.struct(origin=vec3f, raydir=vec3f, raylen=ti.f32)
pixelrays = ray3f.field(shape=(512, 512))
```
Local structs in both python and Taichi scopes can be initialized with either keyword arguments and dictionaries:
```python
ray = ti.Struct(
    origin=ti.Vector([0.0, 0.0, 0.0]), raydir=ti.Vector([0.0, 0.0, 1.0]), raylen=1.0,
)
# or
ray = ti.Struct({
    "origin": ti.Vector([0.0, 0.0, 0.0]),
    "raydir": ti.Vector([0.0, 0.0, 1.0]), 
    "raylen": 1.0,
})
```
In addition to directly invoking initializers of `Struct` and `Matrix` classes, `CompoundType` instances can be also called to create matrix or struct instances. Vectors and matrices can be created using GLSL-like broadcast syntax since the shape of the vector or matrix is already known:
```python
vec2f = ti.types.vector(2, ti.f32)
...
ro = vec3(0.0) # ti.Vector([0.0, 0.0, 0.0])
rd = vec3(vec2(0), 1) # ti.Vector([0.0, 0.0, 1.0]), will perform implicit cast
ray = ray3f(origin=ro, raydir=rd, raylen=1.0)
```

### CODE REVIEW

File: python/taichi/lang/struct.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+    def __setitem__(self, indices, item):
+        raise NotImplementedError(
+            "Cannot assign the whole struct in Python scope")
+
+    def __len__(self):
+        """Get the number of entries in a custom struct"""
+        return len(self.entries)
+
+    def __iter__(self):
+        return self.entries.values()
```
Comment (ljcc0930) [inline python/taichi/lang/struct.py:167]: Keys?

File: python/taichi/lang/struct.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+    def is_global(self):
+        results = [False for _ in self.entries.values()]
+        for i, e in enumerate(self.entries.values()):
+            if isinstance(e, expr.Expr):
+                if e.is_global():
+                    results[i] = True
+            assert results[i] == results[0], \
+                "Structs with mixed global/local entries are not allowed"
+        return results[0]
```
Comment (ljcc0930) [inline python/taichi/lang/struct.py:34]:
```suggestion
    def is_global(self):
        assert len(value) > 0, "Struct has no member!"
        results = [isinstance(_, expr.Expr) and _.is_global()
                       for _ in self.entries.values()]
        assert sum(results) in [0, len(results)], \
            "Structs with mixed global/local entries are not allowed"
        return results[0]
```
nit😂

**Comment by k-ye** [comment]
Ping :-) ?

File: docs/lang/articles/basic/field.md [hunk 3]
```diff
@@ -87,3 +87,33 @@ declare a field of size `64`. E.g., instead of declaring
...
+```python
+# set the position of the first particle to origin
+particle_field[0] # local ti.Struct
+particle_field[0].pos = ti.Vector([0.0, 0.0, 0.0])
+
+# make the mass of all particles be 1
+particle_field.mass # global ti.Vector.field
+particle_field.mass.fill(1.0)
+```
+- See [Structs](../../api/struct.md) for more on matrices.
```
[the remark this answers is not in the document]
Comment (victoriacity) [reply docs/lang/articles/basic/field.md]: The typo has been fixed in the latest commits.

File: tests/python/test_torch_io.py [hunk 4]
```diff
@@ -175,6 +175,25 @@ def test_io_zeros():
     assert zeros[1, 2] == 4
 
 
+@ti.torch_test
```
[the remark this answers is not in the document]
Comment (victoriacity) [reply tests/python/test_torch_io.py]: Fixed in the latest commits.

File: tests/python/test_numpy_io.py [hunk 5]
```diff
@@ -46,6 +46,39 @@ def test_from_numpy_2d():
             assert val[i, j] == i + j * 3
 
 
+@ti.all_archs
```
[the remark this answers is not in the document]
Comment (victoriacity) [reply tests/python/test_numpy_io.py]: Fixed in the latest commits.

File: tests/python/test_numpy_io.py [hunk 6]
```diff
@@ -46,6 +46,39 @@ def test_from_numpy_2d():
...
+        f[i].b = f[i].a * 2
+
+    arr_dict = f.to_numpy()
+
+    for i in range(n):
+        assert arr_dict["a"][i] == i
+        assert arr_dict["b"][i] == i * 2
+
+
+@ti.all_archs
```
[the remark this answers is not in the document]
Comment (victoriacity) [reply tests/python/test_numpy_io.py]: Fixed in the latest commits.

File: python/taichi/lang/struct.py [hunk 7]
```diff
@@ -0,0 +1,549 @@
...
+            if kwargs == {}:
+                raise TaichiSyntaxError(
+                    "Custom type instances need to be created with an initial value."
+                )
+            else:
+                # initialize struct members by keywords
+                entries = kwargs
+        elif len(args) == 1:
+            # fill a single scalar
+            if isinstance(args[0], numbers.Number):
```
[the remark this answers is not in the document]
Comment (victoriacity) [reply python/taichi/lang/struct.py]: Yes, L1391 of matrix.py is the correct implementation, this has been fixed as well.

File: python/taichi/lang/struct.py [hunk 8]
```diff
@@ -0,0 +1,549 @@
...
+        if self.members.keys() != struct.entries.keys():
+            raise TaichiSyntaxError(
+                "Incompatible arguments for custom struct members!")
+        for k, dtype in self.members.items():
+            if isinstance(dtype, CompoundType):
+                struct.entries[k] = dtype.cast(struct.entries[k])
+            else:
+                if in_python_scope():
+                    v = struct.entries[k]
+                    struct.entries[k] = int(
```
[the remark this answers is not in the document]
Comment (victoriacity) [reply python/taichi/lang/struct.py:552]: This is to perform implicit cast when initializing a new local struct with the type constructor in Python scope. The corresponding test has been added as `test_custom_struct/test_compound_type_implicit_cast`; the test passes with the updated code and will fail without the type casting in L536.

File: python/taichi/lang/struct.py [hunk 9]
```diff
@@ -0,0 +1,570 @@
...
+                ret.entries[k] = v.element_wise_unary(foo)
+        return ret
+
+    def element_wise_binary(self, foo, other):
+        _taichi_skip_traceback = 1
+        ret = self.empty_copy()
+        if isinstance(other, (dict)):
+            other = Struct(other)
+        if isinstance(other, Struct):
+            assert self.entries.keys() == other.entries.keys(
```
Comment (strongoier) [inline python/taichi/lang/struct.py:123]: We use `assert` to verify whether the compiler itself is implemented correctly, and `raise` errors for user errors.

File: python/taichi/lang/struct.py [hunk 10]
```diff
@@ -0,0 +1,570 @@
...
+            other = Struct(other)
+        if not isinstance(other, Struct):
+            ret = self.empty_copy()
+            for k, v in ret.items:
+                if isinstance(v, (Matrix, Struct)):
+                    ret.entries[k] = v.broadcast_copy(other)
+                else:
+                    ret.entries[k] = other
+            other = ret
+        assert self.entries.keys() == other.entries.keys(
```
Comment (strongoier) [inline python/taichi/lang/struct.py:150]: ditto

File: python/taichi/lang/struct.py [hunk 11]
```diff
@@ -0,0 +1,570 @@
...
+            other = Struct(other)
+        if is_taichi_class(other):
+            other = other.variable()
+        if foo.__name__ == 'assign' and not isinstance(other, Struct):
+            raise TaichiSyntaxError(
+                'cannot assign scalar expr to '
+                f'taichi class {type(self)}, maybe you want to use `a.fill(b)` instead?'
+            )
+        if isinstance(other, Struct):
+            assert self.entries.keys() == other.entries.keys(
```
Comment (strongoier) [inline python/taichi/lang/struct.py:166]: ditto

**Review by strongoier — COMMENTED** [review]
Would you also resolve conflicts with master?

**Review by strongoier — APPROVED** [review]
LGTM! Thanks for your effort!

**Pull request merged**